crates/holochain/src/core/workflow/validation_receipt_workflow.rs

```diff
@@ -0,0 +1,183 @@
+use super::error::WorkflowError;
+use fallible_iterator::FallibleIterator;
+use holo_hash::DhtOpHash;
+use holochain_cascade::Cascade;
+use holochain_cascade::DbPair;
+use holochain_keystore::KeystoreSender;
+use holochain_lmdb::buffer::KvBufFresh;
+use holochain_lmdb::db::GetDb;
+use holochain_lmdb::db::INTEGRATED_DHT_OPS;
+use holochain_lmdb::env::EnvironmentRead;
+use holochain_lmdb::fresh_reader;
+use holochain_lmdb::prelude::*;
+use holochain_p2p::HolochainP2pCell;
+use holochain_p2p::HolochainP2pCellT;
+use holochain_state::prelude::*;
+use holochain_zome_types::TryInto;
+use tracing::*;
+
+use crate::core::queue_consumer::OneshotWriter;
+use crate::core::queue_consumer::WorkComplete;
+
+use super::error::WorkflowResult;
+
+#[cfg(test)]
+mod tests;
+
+#[instrument(skip(workspace, writer, network))]
+/// Send validation receipts to their authors in serial and wait for
+/// responses. Might not be fast but only requires a single outgoing
+/// connection at any moment.
+/// TODO: If we want faster and better feedback on when our data
+/// has "hit" the network (it's safe to close the laptop lid) then
+/// we could batch this function to the amount of outgoing connections
+/// we are happy with.
+/// Also we could retry on some schedule rather then waiting for a
+/// future op to trigger this workflow.
+/// TODO: We have a bool [`holochain_p2p::HolochainP2pCellT::publish`]
+/// in publishing ops that specifies whether or not
+/// we want a receipt. It is currently ignored here because it can't be set
+/// anywhere. If we do choose to use that bool then we need to use it here.
+pub async fn validation_receipt_workflow(
+    mut workspace: ValidationReceiptWorkspace,
+    writer: OneshotWriter,
+    network: &mut HolochainP2pCell,
+) -> WorkflowResult<WorkComplete> {
+    // Get the env and keystore
+    let env = workspace.elements.headers().env().clone();
+    let keystore = workspace.keystore.clone();
+
+    // Get out all ops that we have not received acknowledgments from the author yet.
+    let ops: Vec<(DhtOpHash, IntegratedDhtOpsValue)> = fresh_reader!(env, |r| workspace
+        .integrated_dht_ops
+        .iter(&r)?
+        .filter(|(_, v)| Ok(!v.receipt_acknowledged))
```

Comment:
@artbrock - I had been thinking of the responsibility for collecting receipts on the author. I think validators should only make one single attempt to send a receipt, and move on... If the author doesn't receive enough receipts, they can publish again, even to validators that have already integrated the op - at which point sending a receipt is easy. Thoughts?

Reply:
Publishing again won't cause receipts to be resent as new dup integrated ops are just dropped.
We'd need to add a request receipt message that authors can send.

Reply:
As an author, you don't know if a particular validating node:
- has integrated your op from gossip
- has integrated your op from previous publish
- is pending validation from either of the above
- has never seen your op

I'm not sure a request_receipt message makes sense if they may not know the data of the op - Hence, I was assuming we could call publish again, and *if* they have already integrated that op for any reason, we could add some code to send a validation receipt.

Reply:
Also, I don't think we want to always publish validation receipts, that would mean validating nodes are always having to establish connections to authors - I think we should only try to send the receipt if we actually got a publish from a node, not in the case of receiving it from gossip.

Reply:
There's one big problem with this. We currently can't tell if it was an op from publish or gossip because kitsune pipes them all through publish

Reply:
I realized that this can be done by checking the from agent

Reply:
Hmm... we might want to make it more explicit, though... technically an author could occasionally be in a neighborhood and end up gossiping their own item : )